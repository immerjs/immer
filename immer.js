"use strict"
// @ts-check

if (typeof Proxy === "undefined")
    throw new Error(
        "Immer requires `Proxy` to be available, but it seems to be not available on your platform. Consider requiring immer '\"immer/es5\"' instead."
    )

const PROXY_STATE = Symbol("immer-proxy-state") // TODO: create per closure, to avoid sharing proxies between multiple immer version
let autoFreeze = true

const objectTraps = {
    get(target, prop) {
        if (prop === PROXY_STATE) return target
        return target.get(prop)
    },
    has(target, prop) {
        return prop in target.source
    },
    ownKeys(target) {
        return Reflect.ownKeys(target.source)
    },
    set(target, prop, value) {
        target.set(prop, value)
        return true
    },
    deleteProperty(target, prop) {
        target.deleteProp(prop)
        return true
    },
    getOwnPropertyDescriptor(target, prop) {
        return target.getOwnPropertyDescriptor(prop)
    },
    defineProperty(target, property, descriptor) {
        target.defineProperty(property, descriptor)
        return true
    },
    setPrototypeOf() {
        throw new Error("Don't even try this...")
    }
}

const arrayTraps = {
    get(target, prop) {
        if (prop === PROXY_STATE) return target[0]
        return target[0].get(prop)
    },
    has(target, prop) {
        return prop in target[0].source
    },
    ownKeys(target) {
        return Reflect.ownKeys(target[0].source)
    },
    set(target, prop, value) {
        target[0].set(prop, value)
        return true
    },
    deleteProperty(target, prop) {
        target[0].deleteProp(prop)
        return true
    },
    getOwnPropertyDescriptor(target, prop) {
        return target[0].getOwnPropertyDescriptor(prop)
    },
    defineProperty(target, property, descriptor) {
        target[0].defineProperty(property, descriptor)
        return true
    },
    setPrototypeOf() {
        throw new Error("Don't even try this...")
    }
}

/**
 * Immer takes a state, and runs a function against it.
 * That function can freely mutate the state, as it will create copies-on-write.
 * This means that the original state will stay unchanged, and once the function finishes, the modified state is returned
 *
 * @export
 * @param {any} baseState - the state to start with
 * @param {Function} thunk - function that receives a proxy of the base state as first argument and which can be freely modified
 * @returns {any} a new state, or the base state if nothing was modified
 */
function immer(baseState, thunk) {
    const revocableProxies = []

    class State {
        constructor(parent, base) {
            this.modified = false
            this.parent = parent
            this.base = base
            this.copy = undefined
            this.proxies = {}
        }

        get source() {
            return this.modified === true ? this.copy : this.base
        }

        get(prop) {
            if (this.modified) {
                const value = this.copy[prop]
                if (!isProxy(value) && isProxyable(value))
                    return (this.copy[prop] = createProxy(this, value))
                return value
            } else {
                if (prop in this.proxies) return this.proxies[prop]
                const value = this.base[prop]
                if (!isProxy(value) && isProxyable(value))
                    return (this.proxies[prop] = createProxy(this, value))
                return value
            }
        }

        set(prop, value) {
            if (!this.modified) {
                if (
                    (prop in this.base && this.base[prop] === value) ||
                    (prop in this.proxies && this.proxies[prop] === value)
                )
                    return
                this.markChanged()
            }
            this.copy[prop] = value
        }

        deleteProp(prop) {
            this.markChanged()
            delete this.copy[prop]
        }

        getOwnPropertyDescriptor(prop) {
            const owner = this.modified
                ? this.copy
                : prop in this.proxies ? this.proxies : this.base
            const descriptor = Reflect.getOwnPropertyDescriptor(owner, prop)
            if (descriptor) descriptor.configurable = true // XXX: is this really needed?
            return descriptor
        }

        defineProperty(property, descriptor) {
            this.markChanged()
            Object.defineProperty(this.copy, property, descriptor)
        }

        markChanged() {
            if (!this.modified) {
                this.modified = true
                this.copy = Array.isArray(this.base) // TODO: eliminate those isArray checks?
                    ? this.base.slice()
                    : Object.assign({}, this.base)
                Object.assign(this.copy, this.proxies) // yup that works for arrays as well
                if (this.parent) this.parent.markChanged()
            }
        }
    }

    // creates a proxy for plain objects / arrays
    function createProxy(parentState, base) {
        const state = new State(parentState, base)
        let proxy
        if (Array.isArray(base)) {
            // Proxy should be created with an array to make it an array for JS
            // so... here you have it!
            proxy = Proxy.revocable([state], arrayTraps)
        } else {
            proxy = Proxy.revocable(state, objectTraps)
        }
        revocableProxies.push(proxy)
        return proxy.proxy
    }

    // given a base object, returns it if unmodified, or return the changed cloned if modified
    function finalize(base) {
        if (isProxy(base)) {
            const state = base[PROXY_STATE]
            if (state.modified === true) {
                if (Array.isArray(state.base)) return finalizeArray(state)
                return finalizeObject(state)
            } else return state.base
        }
        return base
    }

    function finalizeObject(state) {
        const copy = state.copy
        Object.keys(copy).forEach(prop => {
            copy[prop] = finalize(copy[prop])
        })
        return freeze(copy)
    }

    function finalizeArray(state) {
        const copy = state.copy
        copy.forEach((value, index) => {
            copy[index] = finalize(copy[index])
        })
        return freeze(copy)
    }

    // create proxy for root
    const rootClone = createProxy(undefined, baseState)
    // execute the thunk
    const maybeVoidReturn = thunk(rootClone)
    //values either than undefined will trigger warning;
    !Object.is(maybeVoidReturn, undefined) &&
        console.warn(
            `Immer callback expects no return value. However ${typeof maybeVoidReturn} was returned`
        )
    // and finalize the modified proxy
    const res = finalize(rootClone)
    // revoke all proxies
    revocableProxies.forEach(p => p.revoke())
    return res
}

function isProxy(value) {
    return !!value && !!value[PROXY_STATE]
}

function isProxyable(value) {
    if (!value) return false
    if (typeof value !== "object") return false
    if (Array.isArray(value)) return true
    const proto = Object.getPrototypeOf(value)
    return (proto === proto) === null || Object.prototype
}

function freeze(value) {
    // short circuit to achieve 100% code coverage instead of 98%
    /*
    if(autoFreeze) {
        Object.freeze(value)
    }
    * */
    autoFreeze && Object.freeze(value);
    return value
}

/**
 * Automatically freezes any state trees generated by immer.
 * This protects against accidental modifications of the state tree outside of an immer function.
 * This comes with a performance impact, so it is recommended to disable this option in production.
 * It is by default enabled.
 *
 * @returns {void}
 */
function setAutoFreeze(enableAutoFreeze) {
    autoFreeze = enableAutoFreeze
}

Object.defineProperty(exports, "__esModule", {
    value: true
})
module.exports.default = immer
module.exports.setAutoFreeze = setAutoFreeze
