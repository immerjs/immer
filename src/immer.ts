import {
	applyPatchesImpl,
	createDraftImpl,
	createImmerContext,
	Draft,
	finishDraftImpl,
	Immer,
	Immutable,
	IProduce,
	IProduceWithPatches,
	Objectish,
	Patch,
	PatchListener,
	produceImpl,
	produceWithPatchesImpl,
	setAutoFreezeImpl,
	setUseStrictShallowCopyImpl
} from "./internal"

export {
	Draft,
	WritableDraft,
	Immutable,
	Patch,
	PatchListener,
	Producer,
	original,
	current,
	isDraft,
	isDraftable,
	NOTHING as nothing,
	DRAFTABLE as immerable,
	freeze,
	Objectish,
	StrictMode
} from "./internal"

const globalImmerContext = createImmerContext()

/**
 * The `produce` function takes a value and a "recipe function" (whose
 * return value often depends on the base state). The recipe function is
 * free to mutate its first argument however it wants. All mutations are
 * only ever applied to a __copy__ of the base state.
 *
 * Pass only a function to create a "curried producer" which relieves you
 * from passing the recipe function every time.
 *
 * Only plain objects and arrays are made mutable. All other objects are
 * considered uncopyable.
 *
 * Note: This function is __bound__ to its `Immer` instance.
 *
 * @param {any} base - the initial state
 * @param {Function} producer - function that receives a proxy of the base state as first argument and which can be freely modified
 * @param {Function} patchListener - optional function that will be called with all the patches produced here
 * @returns {any} a new state, or the initial state if nothing was modified
 */
export const produce: IProduce = produceImpl.bind(globalImmerContext)

/**
 * Like `produce`, but `produceWithPatches` always returns a tuple
 * [nextState, patches, inversePatches] (instead of just the next state)
 */
export const produceWithPatches: IProduceWithPatches = produceWithPatchesImpl.bind(
	globalImmerContext
)

/**
 * Pass true to automatically freeze all copies created by Immer.
 *
 * Always freeze by default, even in production mode
 */
export const setAutoFreeze = setAutoFreezeImpl.bind(globalImmerContext)

/**
 * Pass true to enable strict shallow copy.
 *
 * By default, immer does not copy the object descriptors such as getter, setter and non-enumrable properties.
 */
export const setUseStrictShallowCopy = setUseStrictShallowCopyImpl.bind(
	globalImmerContext
)

/**
 * Apply an array of Immer patches to the first argument.
 *
 * This function is a producer, which means copy-on-write is in effect.
 */
export const applyPatches = applyPatchesImpl.bind(globalImmerContext) as <
	T extends Objectish
>(
	base: T,
	patches: readonly Patch[]
) => T

/**
 * Create an Immer draft from the given base state, which may be a draft itself.
 * The draft can be modified until you finalize it with the `finishDraft` function.
 */
export const createDraft = createDraftImpl.bind(globalImmerContext) as <
	T extends Objectish
>(
	base: T
) => Draft<T>

/**
 * Finalize an Immer draft from a `createDraft` call, returning the base state
 * (if no changes were made) or a modified copy. The draft must *not* be
 * mutated afterwards.
 *
 * Pass a function as the 2nd argument to generate Immer patches based on the
 * changes that were made.
 */
export const finishDraft = finishDraftImpl.bind(globalImmerContext) as <
	D extends Draft<any>
>(
	draft: D,
	patchListener?: PatchListener
) => D extends Draft<infer T> ? T : never

/**
 * This function is actually a no-op, but can be used to cast an immutable type
 * to an draft type and make TypeScript happy
 *
 * @param value
 */
export function castDraft<T>(value: T): Draft<T> {
	return value as any
}

/**
 * This function is actually a no-op, but can be used to cast a mutable type
 * to an immutable type and make TypeScript happy
 * @param value
 */
export function castImmutable<T>(value: T): Immutable<T> {
	return value as any
}

export {Immer}

export {enablePatches} from "./plugins/patches"
export {enableMapSet} from "./plugins/mapset"
