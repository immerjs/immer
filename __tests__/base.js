import immer from ".."

describe("base", () => {
  let baseState
  let origBaseState

  beforeEach(() => {
    origBaseState = baseState = createBaseState()
  })

  it("should return the original without modifications", () => {
    const nextState = immer(baseState, () => {})
    expect(nextState).toBe(baseState)
  })

  it("should return the original without modifications when reading stuff", () => {
    const nextState = immer(baseState, s => {
      expect(s.aProp).toBe("hi")
      expect(s.anObject.nested).toEqual({yummie: true})
    })
    expect(nextState).toBe(baseState)
  })

  it("should not return any value: thunk", () => {
    const warning = jest.spyOn(console, "warn")
    immer(baseState, () => ({bad: "don't do this"}))
    immer(baseState, () => [1, 2, 3])
    immer(baseState, () => false)
    immer(baseState, () => "")

    expect(warning).toHaveBeenCalledTimes(4)
  })

  it("should return a copy when modifying stuff", () => {
    const nextState = immer(baseState, s => {
      s.aProp = "hello world"
    })
    expect(nextState).not.toBe(baseState)
    expect(baseState.aProp).toBe("hi")
    expect(nextState.aProp).toBe("hello world")
    // structural sharing?
    expect(nextState.nested).toBe(baseState.nested)
  })

  it("deep change bubbles up", () => {
    const nextState = immer(baseState, s => {
      s.anObject.nested.yummie = false
    })
    expect(nextState).not.toBe(baseState)
    expect(nextState.anObject).not.toBe(baseState.anObject)
    expect(baseState.anObject.nested.yummie).toBe(true)
    expect(nextState.anObject.nested.yummie).toBe(false)
    expect(nextState.anArray).toBe(baseState.anArray)
  })

  it("can add props", () => {
    const nextState = immer(baseState, s => {
      s.anObject.cookie = {tasty: true}
    })
    expect(nextState).not.toBe(baseState)
    expect(nextState.anObject).not.toBe(baseState.anObject)
    expect(nextState.anObject.nested).toBe(baseState.anObject.nested)
    expect(nextState.anObject.cookie).toEqual({tasty: true})
  })

  it("can delete props", () => {
    const nextState = immer(baseState, s => {
      delete s.anObject.nested
    })
    expect(nextState).not.toBe(baseState)
    expect(nextState.anObject).not.toBe(baseState.anObject)
    expect(nextState.anObject.nested).toBe(undefined)
  })

  it("ignores single non-modification", () => {
    const nextState = immer(baseState, s => {
      s.aProp = "hi"
    })
    expect(nextState).toBe(baseState)
  })

  it("processes single modification", () => {
    const nextState = immer(baseState, s => {
      s.aProp = "hello"
      s.aProp = "hi"
    })
    expect(nextState).not.toBe(baseState)
    expect(nextState).toEqual(baseState)
  })

  it("should support reading arrays", () => {
    const nextState = immer(baseState, s => {
      s.anArray.slice()
    })
    expect(nextState.anArray).toBe(baseState.anArray)
    expect(nextState).toBe(baseState)
  })

  it("should support changing arrays", () => {
    const nextState = immer(baseState, s => {
      s.anArray[3] = true
    })
    expect(nextState).not.toBe(baseState)
    expect(nextState.anArray).not.toBe(baseState.anArray)
    expect(nextState.anArray[3]).toEqual(true)
  })

  it("should support changing arrays - 2", () => {
    const nextState = immer(baseState, s => {
      s.anArray.splice(1, 1, "a", "b")
    })
    expect(nextState).not.toBe(baseState)
    expect(nextState.anArray).not.toBe(baseState.anArray)

    expect(nextState.anArray).toEqual([3, "a", "b", {c: 3}, 1])
  })

  it("should support sorting arrays", () => {
    const nextState = immer(baseState, s => {
      s.anArray[2].c = 4
      s.anArray.sort()
      s.anArray[3].c = 5
    })
    expect(nextState).not.toBe(baseState)
    expect(nextState.anArray).not.toBe(baseState.anArray)
    expect(nextState.anArray).toEqual([1, 2, 3, {c: 5}])
  })

  it("should updating inside arrays", () => {
    const nextState = immer(baseState, s => {
      s.anArray[2].test = true
    })
    expect(nextState).not.toBe(baseState)
    expect(nextState.anArray).not.toBe(baseState.anArray)
    expect(nextState.anArray).toEqual([3, 2, {c: 3, test: true}, 1])
  })

  it("reusing object should work", () => {
    const nextState = immer(baseState, s => {
      debugger
      const obj = s.anObject
      delete s.anObject
      s.messy = obj
    })
    expect(nextState).not.toBe(baseState)
    expect(nextState.anArray).toBe(baseState.anArray)
    expect(nextState).toEqual({
      anArray: [3, 2, {c: 3}, 1],
      aProp: "hi",
      messy: {
        nested: {
          yummie: true,
        },
        coffee: false,
      },
    })
    expect(nextState.messy.nested).toBe(baseState.anObject.nested)
  })

  it("refs should be transparent", () => {
    const nextState = immer(baseState, s => {
      const obj = s.anObject
      s.aProp = "hello"
      delete s.anObject
      obj.coffee = true
      s.messy = obj
    })
    expect(nextState).not.toBe(baseState)
    expect(nextState.anArray).toBe(baseState.anArray)
    expect(nextState).toEqual({
      anArray: [3, 2, {c: 3}, 1],
      aProp: "hello",
      messy: {
        nested: {
          yummie: true,
        },
        coffee: true,
      },
    })
    expect(nextState.messy.nested).toBe(baseState.anObject.nested)
  })

  it("should allow setting to undefined a defined draft property", () => {
    const nextState = immer(baseState, s => {
      s.aProp = undefined
    })
    expect(nextState).not.toBe(baseState)
    expect(baseState.aProp).toBe("hi")
    expect(nextState.aProp).toBe(undefined)
  })

  afterEach(() => {
    expect(baseState).toBe(origBaseState)
    expect(baseState).toEqual(createBaseState())
  })

  function createBaseState() {
    return {
      anArray: [3, 2, {c: 3}, 1],
      aProp: "hi",
      anObject: {
        nested: {
          yummie: true,
        },
        coffee: false,
      },
    }
  }
})

describe("readme example", () => {
  it("works", () => {
    const baseState = [
      {
        todo: "Learn typescript",
        done: true,
      },
      {
        todo: "Try immer",
        done: false,
      },
    ]

    const nextState = immer(baseState, state => {
      state.push({todo: "Tweet about it"})
      state[1].done = true
    })

    // the new item is only added to the next state,
    // base state is unmodified
    expect(baseState.length).toBe(2)
    expect(nextState.length).toBe(3)

    // same for the changed 'done' prop
    expect(baseState[1].done).toBe(false)
    expect(nextState[1].done).toBe(true)

    // unchanged data is structurally shared
    expect(nextState[0]).toBe(baseState[0])
    // changed data not (d√ªh)
    expect(nextState[1]).not.toBe(baseState[1])
  })
})
