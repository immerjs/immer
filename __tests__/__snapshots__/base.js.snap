// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base functionality - es5 (autofreeze) async recipe function works with rejected promises 1`] = `"Cannot use a proxy that has been revoked. Did you pass an object from inside an immer function to an async process? {\\"a\\":0,\\"b\\":1}"`;

exports[`base functionality - es5 (autofreeze) recipe functions cannot return a modified child draft 1`] = `"An immer producer returned a new value *and* modified its draft. Either return a new value *or* modify the draft."`;

exports[`base functionality - es5 (autofreeze) recipe functions cannot return an object that references itself 1`] = `"Immer forbids circular references"`;

exports[`base functionality - es5 (autofreeze) revokes the draft once produce returns 1`] = `"Cannot use a proxy that has been revoked. Did you pass an object from inside an immer function to an async process? {\\"a\\":1}"`;

exports[`base functionality - es5 (autofreeze) revokes the draft once produce returns 2`] = `"Cannot use a proxy that has been revoked. Did you pass an object from inside an immer function to an async process? {\\"a\\":1}"`;

exports[`base functionality - es5 (autofreeze) revokes the draft once produce returns 3`] = `"Cannot use a proxy that has been revoked. Did you pass an object from inside an immer function to an async process? [1]"`;

exports[`base functionality - es5 (autofreeze) revokes the draft once produce returns 4`] = `"Cannot use a proxy that has been revoked. Did you pass an object from inside an immer function to an async process? [1]"`;

exports[`base functionality - es5 (autofreeze) throws on computed properties 1`] = `"Immer drafts cannot have computed properties"`;

exports[`base functionality - es5 (autofreeze) throws when the draft is modified and another object is returned 1`] = `"An immer producer returned a new value *and* modified its draft. Either return a new value *or* modify the draft."`;

exports[`base functionality - es5 (autofreeze)(patch listener) async recipe function works with rejected promises 1`] = `"Cannot use a proxy that has been revoked. Did you pass an object from inside an immer function to an async process? {\\"a\\":0,\\"b\\":1}"`;

exports[`base functionality - es5 (autofreeze)(patch listener) recipe functions cannot return a modified child draft 1`] = `"An immer producer returned a new value *and* modified its draft. Either return a new value *or* modify the draft."`;

exports[`base functionality - es5 (autofreeze)(patch listener) recipe functions cannot return an object that references itself 1`] = `"Immer forbids circular references"`;

exports[`base functionality - es5 (autofreeze)(patch listener) revokes the draft once produce returns 1`] = `"Cannot use a proxy that has been revoked. Did you pass an object from inside an immer function to an async process? {\\"a\\":1}"`;

exports[`base functionality - es5 (autofreeze)(patch listener) revokes the draft once produce returns 2`] = `"Cannot use a proxy that has been revoked. Did you pass an object from inside an immer function to an async process? {\\"a\\":1}"`;

exports[`base functionality - es5 (autofreeze)(patch listener) revokes the draft once produce returns 3`] = `"Cannot use a proxy that has been revoked. Did you pass an object from inside an immer function to an async process? [1]"`;

exports[`base functionality - es5 (autofreeze)(patch listener) revokes the draft once produce returns 4`] = `"Cannot use a proxy that has been revoked. Did you pass an object from inside an immer function to an async process? [1]"`;

exports[`base functionality - es5 (autofreeze)(patch listener) throws on computed properties 1`] = `"Immer drafts cannot have computed properties"`;

exports[`base functionality - es5 (autofreeze)(patch listener) throws when the draft is modified and another object is returned 1`] = `"An immer producer returned a new value *and* modified its draft. Either return a new value *or* modify the draft."`;

exports[`base functionality - es5 (no freeze) async recipe function works with rejected promises 1`] = `"Cannot use a proxy that has been revoked. Did you pass an object from inside an immer function to an async process? {\\"a\\":0,\\"b\\":1}"`;

exports[`base functionality - es5 (no freeze) recipe functions cannot return a modified child draft 1`] = `"An immer producer returned a new value *and* modified its draft. Either return a new value *or* modify the draft."`;

exports[`base functionality - es5 (no freeze) recipe functions cannot return an object that references itself 1`] = `"Immer forbids circular references"`;

exports[`base functionality - es5 (no freeze) revokes the draft once produce returns 1`] = `"Cannot use a proxy that has been revoked. Did you pass an object from inside an immer function to an async process? {\\"a\\":1}"`;

exports[`base functionality - es5 (no freeze) revokes the draft once produce returns 2`] = `"Cannot use a proxy that has been revoked. Did you pass an object from inside an immer function to an async process? {\\"a\\":1}"`;

exports[`base functionality - es5 (no freeze) revokes the draft once produce returns 3`] = `"Cannot use a proxy that has been revoked. Did you pass an object from inside an immer function to an async process? [1]"`;

exports[`base functionality - es5 (no freeze) revokes the draft once produce returns 4`] = `"Cannot use a proxy that has been revoked. Did you pass an object from inside an immer function to an async process? [1]"`;

exports[`base functionality - es5 (no freeze) throws on computed properties 1`] = `"Immer drafts cannot have computed properties"`;

exports[`base functionality - es5 (no freeze) throws when the draft is modified and another object is returned 1`] = `"An immer producer returned a new value *and* modified its draft. Either return a new value *or* modify the draft."`;

exports[`base functionality - proxy (autofreeze) array drafts throws when a non-numeric property is added 1`] = `"Immer only supports setting array indices and the 'length' property"`;

exports[`base functionality - proxy (autofreeze) array drafts throws when a non-numeric property is deleted 1`] = `"Immer only supports deleting array indices"`;

exports[`base functionality - proxy (autofreeze) async recipe function works with rejected promises 1`] = `"Cannot perform 'get' on a proxy that has been revoked"`;

exports[`base functionality - proxy (autofreeze) recipe functions cannot return a modified child draft 1`] = `"An immer producer returned a new value *and* modified its draft. Either return a new value *or* modify the draft."`;

exports[`base functionality - proxy (autofreeze) recipe functions cannot return an object that references itself 1`] = `"Immer forbids circular references"`;

exports[`base functionality - proxy (autofreeze) revokes the draft once produce returns 1`] = `"Cannot perform 'get' on a proxy that has been revoked"`;

exports[`base functionality - proxy (autofreeze) revokes the draft once produce returns 2`] = `"Cannot perform 'set' on a proxy that has been revoked"`;

exports[`base functionality - proxy (autofreeze) revokes the draft once produce returns 3`] = `"Cannot perform 'get' on a proxy that has been revoked"`;

exports[`base functionality - proxy (autofreeze) revokes the draft once produce returns 4`] = `"Cannot perform 'set' on a proxy that has been revoked"`;

exports[`base functionality - proxy (autofreeze) revokes the draft once produce returns 5`] = `"Cannot perform 'get' on a proxy that has been revoked"`;

exports[`base functionality - proxy (autofreeze) revokes the draft once produce returns 6`] = `"Cannot perform 'set' on a proxy that has been revoked"`;

exports[`base functionality - proxy (autofreeze) revokes the draft once produce returns 7`] = `"Cannot perform 'get' on a proxy that has been revoked"`;

exports[`base functionality - proxy (autofreeze) revokes the draft once produce returns 8`] = `"Cannot perform 'set' on a proxy that has been revoked"`;

exports[`base functionality - proxy (autofreeze) throws on computed properties 1`] = `"Immer drafts cannot have computed properties"`;

exports[`base functionality - proxy (autofreeze) throws when Object.defineProperty() is used on drafts 1`] = `"Object.defineProperty() cannot be used on an Immer draft"`;

exports[`base functionality - proxy (autofreeze) throws when Object.setPrototypeOf() is used on a draft 1`] = `"Object.setPrototypeOf() cannot be used on an Immer draft"`;

exports[`base functionality - proxy (autofreeze) throws when the draft is modified and another object is returned 1`] = `"An immer producer returned a new value *and* modified its draft. Either return a new value *or* modify the draft."`;

exports[`base functionality - proxy (autofreeze)(patch listener) array drafts throws when a non-numeric property is added 1`] = `"Immer only supports setting array indices and the 'length' property"`;

exports[`base functionality - proxy (autofreeze)(patch listener) array drafts throws when a non-numeric property is deleted 1`] = `"Immer only supports deleting array indices"`;

exports[`base functionality - proxy (autofreeze)(patch listener) async recipe function works with rejected promises 1`] = `"Cannot perform 'get' on a proxy that has been revoked"`;

exports[`base functionality - proxy (autofreeze)(patch listener) recipe functions cannot return a modified child draft 1`] = `"An immer producer returned a new value *and* modified its draft. Either return a new value *or* modify the draft."`;

exports[`base functionality - proxy (autofreeze)(patch listener) recipe functions cannot return an object that references itself 1`] = `"Immer forbids circular references"`;

exports[`base functionality - proxy (autofreeze)(patch listener) revokes the draft once produce returns 1`] = `"Cannot perform 'get' on a proxy that has been revoked"`;

exports[`base functionality - proxy (autofreeze)(patch listener) revokes the draft once produce returns 2`] = `"Cannot perform 'set' on a proxy that has been revoked"`;

exports[`base functionality - proxy (autofreeze)(patch listener) revokes the draft once produce returns 3`] = `"Cannot perform 'get' on a proxy that has been revoked"`;

exports[`base functionality - proxy (autofreeze)(patch listener) revokes the draft once produce returns 4`] = `"Cannot perform 'set' on a proxy that has been revoked"`;

exports[`base functionality - proxy (autofreeze)(patch listener) revokes the draft once produce returns 5`] = `"Cannot perform 'get' on a proxy that has been revoked"`;

exports[`base functionality - proxy (autofreeze)(patch listener) revokes the draft once produce returns 6`] = `"Cannot perform 'set' on a proxy that has been revoked"`;

exports[`base functionality - proxy (autofreeze)(patch listener) revokes the draft once produce returns 7`] = `"Cannot perform 'get' on a proxy that has been revoked"`;

exports[`base functionality - proxy (autofreeze)(patch listener) revokes the draft once produce returns 8`] = `"Cannot perform 'set' on a proxy that has been revoked"`;

exports[`base functionality - proxy (autofreeze)(patch listener) throws on computed properties 1`] = `"Immer drafts cannot have computed properties"`;

exports[`base functionality - proxy (autofreeze)(patch listener) throws when Object.defineProperty() is used on drafts 1`] = `"Object.defineProperty() cannot be used on an Immer draft"`;

exports[`base functionality - proxy (autofreeze)(patch listener) throws when Object.setPrototypeOf() is used on a draft 1`] = `"Object.setPrototypeOf() cannot be used on an Immer draft"`;

exports[`base functionality - proxy (autofreeze)(patch listener) throws when the draft is modified and another object is returned 1`] = `"An immer producer returned a new value *and* modified its draft. Either return a new value *or* modify the draft."`;

exports[`base functionality - proxy (no freeze) array drafts throws when a non-numeric property is added 1`] = `"Immer only supports setting array indices and the 'length' property"`;

exports[`base functionality - proxy (no freeze) array drafts throws when a non-numeric property is deleted 1`] = `"Immer only supports deleting array indices"`;

exports[`base functionality - proxy (no freeze) async recipe function works with rejected promises 1`] = `"Cannot perform 'get' on a proxy that has been revoked"`;

exports[`base functionality - proxy (no freeze) recipe functions cannot return a modified child draft 1`] = `"An immer producer returned a new value *and* modified its draft. Either return a new value *or* modify the draft."`;

exports[`base functionality - proxy (no freeze) recipe functions cannot return an object that references itself 1`] = `"Immer forbids circular references"`;

exports[`base functionality - proxy (no freeze) revokes the draft once produce returns 1`] = `"Cannot perform 'get' on a proxy that has been revoked"`;

exports[`base functionality - proxy (no freeze) revokes the draft once produce returns 2`] = `"Cannot perform 'set' on a proxy that has been revoked"`;

exports[`base functionality - proxy (no freeze) revokes the draft once produce returns 3`] = `"Cannot perform 'get' on a proxy that has been revoked"`;

exports[`base functionality - proxy (no freeze) revokes the draft once produce returns 4`] = `"Cannot perform 'set' on a proxy that has been revoked"`;

exports[`base functionality - proxy (no freeze) revokes the draft once produce returns 5`] = `"Cannot perform 'get' on a proxy that has been revoked"`;

exports[`base functionality - proxy (no freeze) revokes the draft once produce returns 6`] = `"Cannot perform 'set' on a proxy that has been revoked"`;

exports[`base functionality - proxy (no freeze) revokes the draft once produce returns 7`] = `"Cannot perform 'get' on a proxy that has been revoked"`;

exports[`base functionality - proxy (no freeze) revokes the draft once produce returns 8`] = `"Cannot perform 'set' on a proxy that has been revoked"`;

exports[`base functionality - proxy (no freeze) throws on computed properties 1`] = `"Immer drafts cannot have computed properties"`;

exports[`base functionality - proxy (no freeze) throws when Object.defineProperty() is used on drafts 1`] = `"Object.defineProperty() cannot be used on an Immer draft"`;

exports[`base functionality - proxy (no freeze) throws when Object.setPrototypeOf() is used on a draft 1`] = `"Object.setPrototypeOf() cannot be used on an Immer draft"`;

exports[`base functionality - proxy (no freeze) throws when the draft is modified and another object is returned 1`] = `"An immer producer returned a new value *and* modified its draft. Either return a new value *or* modify the draft."`;

exports[`complex nesting map / set / object modify deep object 1`] = `
Object {
  "map": Map {
    "set1" => Set {
      Object {
        "a": 2,
      },
      Object {
        "b": 2,
      },
    },
    "set2" => Set {
      Object {
        "c": 3,
      },
    },
  },
}
`;

exports[`complex nesting map / set / object modify deep object 2`] = `
Array [
  Object {
    "op": "remove",
    "path": Array [
      "map",
      "set1",
      0,
    ],
    "value": Object {
      "a": 1,
    },
  },
  Object {
    "op": "add",
    "path": Array [
      "map",
      "set1",
      0,
    ],
    "value": Object {
      "a": 2,
    },
  },
]
`;

exports[`complex nesting map / set / object modify deep object 3`] = `
Object {
  "map": Map {
    "set1" => Set {
      Object {
        "a": 2,
      },
      Object {
        "b": 2,
      },
    },
    "set2" => Set {
      Object {
        "c": 3,
      },
    },
  },
}
`;

exports[`complex nesting map / set / object modify deep object 4`] = `
Array [
  Object {
    "op": "remove",
    "path": Array [
      "map",
      "set1",
      0,
    ],
    "value": Object {
      "a": 1,
    },
  },
  Object {
    "op": "add",
    "path": Array [
      "map",
      "set1",
      0,
    ],
    "value": Object {
      "a": 2,
    },
  },
]
`;

exports[`complex nesting map / set / object modify deep object 5`] = `
Object {
  "map": Map {
    "set1" => Set {
      Object {
        "a": 2,
      },
      Object {
        "b": 2,
      },
    },
    "set2" => Set {
      Object {
        "c": 3,
      },
    },
  },
}
`;

exports[`complex nesting map / set / object modify deep object 6`] = `
Array [
  Object {
    "op": "remove",
    "path": Array [
      "map",
      "set1",
      0,
    ],
    "value": Object {
      "a": 1,
    },
  },
  Object {
    "op": "add",
    "path": Array [
      "map",
      "set1",
      0,
    ],
    "value": Object {
      "a": 2,
    },
  },
]
`;

exports[`complex nesting map / set / object modify deep object 7`] = `
Object {
  "map": Map {
    "set1" => Set {
      Object {
        "a": 2,
      },
      Object {
        "b": 2,
      },
    },
    "set2" => Set {
      Object {
        "c": 3,
      },
    },
  },
}
`;

exports[`complex nesting map / set / object modify deep object 8`] = `
Array [
  Object {
    "op": "remove",
    "path": Array [
      "map",
      "set1",
      0,
    ],
    "value": Object {
      "a": 1,
    },
  },
  Object {
    "op": "add",
    "path": Array [
      "map",
      "set1",
      0,
    ],
    "value": Object {
      "a": 2,
    },
  },
]
`;

exports[`complex nesting map / set / object modify deep object 9`] = `
Object {
  "map": Map {
    "set1" => Set {
      Object {
        "a": 2,
      },
      Object {
        "b": 2,
      },
    },
    "set2" => Set {
      Object {
        "c": 3,
      },
    },
  },
}
`;

exports[`complex nesting map / set / object modify deep object 10`] = `
Array [
  Object {
    "op": "remove",
    "path": Array [
      "map",
      "set1",
      0,
    ],
    "value": Object {
      "a": 1,
    },
  },
  Object {
    "op": "add",
    "path": Array [
      "map",
      "set1",
      0,
    ],
    "value": Object {
      "a": 2,
    },
  },
]
`;

exports[`complex nesting map / set / object modify deep object 11`] = `
Object {
  "map": Map {
    "set1" => Set {
      Object {
        "a": 2,
      },
      Object {
        "b": 2,
      },
    },
    "set2" => Set {
      Object {
        "c": 3,
      },
    },
  },
}
`;

exports[`complex nesting map / set / object modify deep object 12`] = `
Array [
  Object {
    "op": "remove",
    "path": Array [
      "map",
      "set1",
      0,
    ],
    "value": Object {
      "a": 1,
    },
  },
  Object {
    "op": "add",
    "path": Array [
      "map",
      "set1",
      0,
    ],
    "value": Object {
      "a": 2,
    },
  },
]
`;
